Test/RadixSort/main.cpp

```diff
@@ -38,131 +39,166 @@ class Stopwatch
+	{
 		u32 s;
-		m_sort->configure( m_device, s );
-//		m_sort->setFlag( Oro::RadixSort::FLAG_LOG );
+		m_sort.configure( m_device, s );
+		//		m_sort->setFlag( Oro::RadixSort::FLAG_LOG );
 		OrochiUtils::malloc( m_tempBuffer, s );
 	}
 
-	~SortTest() 
-	{
-		delete m_sort;
-		OrochiUtils::free( m_tempBuffer );
-	}
+	~SortTest() { OrochiUtils::free( m_tempBuffer ); }
 
 	void test( int testSize, const int testBits = 32, const int nRuns = 1 )
 	{
-		using namespace std;
+
 		srand( 123 );
-		vector<u32> src( testSize );
+		std::vector<u32> srcKey( testSize );
 		for( int i = 0; i < testSize; i++ )
 		{
-			src[i] = getRandom( 0u, (u32)(( 1ull << (u64)testBits ) - 1) );
+			srcKey[i] = getRandom( 0u, (u32)( ( 1ull << (u64)testBits ) - 1 ) );
 		}
 
-		u32* srcGpu;
-		u32* dstGpu;
-		OrochiUtils::malloc( srcGpu, testSize );
-		OrochiUtils::malloc( dstGpu, testSize );
+		std::vector<u32> srcValue( testSize );
+		for( int i = 0; i < testSize; i++ )
+		{
+			srcValue[i] = getRandom( 0u, (u32)( ( 1ull << (u64)testBits ) - 1 ) );
+		}
+
+		Oro::RadixSort::KeyValueSoA srcGpu{};
+		Oro::RadixSort::KeyValueSoA dstGpu{};
+
+		OrochiUtils::malloc( srcGpu.key, testSize );
+		OrochiUtils::malloc( dstGpu.key, testSize );
+
+		OrochiUtils::malloc( srcGpu.value, testSize );
+		OrochiUtils::malloc( dstGpu.value, testSize );
 
 		Stopwatch sw;
-		for( int i = 0; i < nRuns ; i++)
+		for( int i = 0; i < nRuns; i++ )
 		{
-			OrochiUtils::copyHtoD( srcGpu, src.data(), testSize );
+			OrochiUtils::copyHtoD( srcGpu.key, srcKey.data(), testSize );
+			OrochiUtils::waitForCompletion();
+
+			OrochiUtils::copyHtoD( srcGpu.value, srcValue.data(), testSize );
 			OrochiUtils::waitForCompletion();
+
 			sw.start();
-			m_sort->sort( srcGpu, dstGpu, testSize, 0, testBits, m_tempBuffer );
+			m_sort.sort( srcGpu, dstGpu, testSize, 0, testBits, m_tempBuffer );
 			OrochiUtils::waitForCompletion();
 			sw.stop();
 			float ms = sw.getMs();
-			float gKeys_s = testSize/1000.f/1000.f/ms;
-			printf("%3.2fms (%3.2fGKeys/s) sorting %3.1fMkeys\n", ms, gKeys_s, testSize/1000.f/1000.f);
+			float gKeys_s = testSize / 1000.f / 1000.f / ms;
+			printf( "%3.2fms (%3.2fGKeys/s) sorting %3.1fMkeys\n", ms, gKeys_s, testSize / 1000.f / 1000.f );
 		}
 
-		vector<u32> dst( testSize );
-		OrochiUtils::copyDtoH( dst.data(), dstGpu, testSize );
+		std::vector<u32> dstKey( testSize );
+		OrochiUtils::copyDtoH( dstKey.data(), dstGpu.key, testSize );
+
+		std::vector<u32> dstValue( testSize );
+		OrochiUtils::copyDtoH( dstValue.data(), dstGpu.value, testSize );
+
+		std::vector<u32> indexHelper( testSize );
+		std::iota( std::begin( indexHelper ), std::end( indexHelper ), 0U );
+
+		std::stable_sort( std::begin( indexHelper ), std::end( indexHelper ), [&]( const auto indexA, const auto indexB ) noexcept { return srcKey[indexA] < srcKey[indexB]; } );
+
+		const auto rearrange = []( auto& targetBuffer, const auto& indexBuffer ) noexcept
```

Comment:
This lambda is used to sort the keys and values according to the content of the helper index buffer.